### PR TITLE
Extend memory efficient attention coverage in Attention/MHA cuda op

### PR DESCRIPTION
**Pull request by wangyems**
### Description

1. upgrade cutlass to 3.0 that containing attn_bias support.
2. extend Attention/MHA to use memory efficient attention when rel_pos_bias with [1, num_head, s, s*] and 1d mask with [2 * batch_size + 1] are present.

new mask format introduction:
MASK_1D_KEY_SEQ_LEN_START,  
[3 * batch_size + 2] with [key_len[0], ..., key_len[batch_size - 1], query_start[0], ..., query_start[batch_size - 1], query_end[batch_size - 1], key_start[0], ..., key_start[batch_size - 1], key_end[batch_size - 1]]

e.g
2D mask with [[1, 1, 1, 0, 0, 0], [1, 1, 1, 1, 1, 0]] converts to this 1D mask is [3, 5, 0, 6, 12, 0, 6, 12]

### Motivation and Context

It potentially benefits tnlrv6 and t5(encoder)

### CODE REVIEW

File: onnxruntime/test/contrib_ops/multihead_attention_op_test.cc [hunk 1]
```diff
@@ -55,7 +55,8 @@ static void RunMultiHeadAttentionTest(
     std::vector<int64_t> key_dims =  {batch_size, is_static_kv ? kv_sequence_length : sequence_length, hidden_size};
     std::vector<int64_t> value_dims = {batch_size, is_static_kv ? kv_sequence_length : sequence_length, v_hidden_size};
     std::vector<int64_t> bias_dims = {hidden_size + hidden_size + v_hidden_size};
-    std::vector<int64_t> rel_pos_bias_dims = {1, num_heads, sequence_length, sequence_length + kv_sequence_length};
+    std::vector<int64_t> rel_pos_bias_dims =
+                         {1, num_heads, sequence_length, past_key_data.size() ? sequence_length + kv_sequence_length : sequence_length};
```
Comment (tianleiwu) [inline onnxruntime/test/contrib_ops/multihead_attention_op_test.cc:59]: past_sequence_length + kv_sequence_length

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.cu [hunk 2]
```diff
@@ -790,8 +803,6 @@ Status QkvToContext(
   float zero = 0.f;
 
   // For raw attention mask, the scalar 1/sqrt(H) is moved to combine with softmax computation.
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/bert/attention_impl.cu:805]: Move this comment with code.

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.cu [hunk 3]
```diff
@@ -754,23 +764,26 @@ Status QkvToContext(
...
     p.num_heads = parameters.num_heads;
     p.sequence_length = parameters.sequence_length;
     p.kv_sequence_length = parameters.total_sequence_length;
     p.qk_head_size = parameters.head_size;
     p.v_head_size = parameters.v_head_size;
     p.causal = parameters.is_unidirectional;
-    p.cu_seqlens_q = nullptr;
-    p.cu_seqlens_k = nullptr;
+    p.scale = scale;
+    p.cu_seqlens_q = nullptr == data.mask_index ? nullptr : const_cast<int32_t*>(reinterpret_cast<const int32_t*>(data.mask_index + batch_size));
```
[the remark this answers is not in the document]
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/bert/attention_impl.cu:775]: added a comment that this design only supports self attention

[the remark this answers is not in the document]
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/bert/attention_impl.cu:775]: l'l support cross attn in this change. let me change the shape from [2*B + 1] to [3*B + 2]

File: onnxruntime/contrib_ops/cuda/bert/cutlass_fmha/kernel_forward.h [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /***************************************************************************************************
- * Copyright (c) 2017 - 2022 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * Copyright (c) 2017 - 2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
[the remark this answers is not in the document]
Comment (wangyems) [reply onnxruntime/contrib_ops/cuda/bert/cutlass_fmha/kernel_forward.h]: removed it

**Review by tianleiwu — APPROVED** [review]

**Review by jchen351 — APPROVED** [review]
LGTM on engineering system side.

**Review by pranavsharma — COMMENTED** [review]
Since we're upgrading cutlass to v3.0 does this change any runtime dep requirements for our users? if yes, we need to document them.

**Comment by wangyems** [comment]
> Since we're upgrading cutlass to v3.0 does this change any runtime dep requirements for our users? if yes, we need to document them.

there's not

**Review by pranavsharma — APPROVED** [review]
Approving for admin.

**Pull request merged**